Ability to setup auto-updates for SW/AppCache

domderen opened the pull request:
SW/AppCache were only updating on page reloads. This PR adds the functionality to have automatic checks for updates to the ServiceWorker/AppCache.

This pull request adds two new properties to the config object:
- `autoUpdate` - if set to true, it initializes a periodic function that checks for updates to the SW/AppCache,
- `autoUpdateInterval` - defines how often the app should check for updates to ServiceWorker/AppCache. Defaults to 1 hour.

Wasn't sure how to compile the code, as Babel was not in the dependencies list, so I've added some setup to make it work, but would appreciate any comments on that part.

domderen commented:
Great, thanks! 

If you can let me know what should I change in terms of the build process, I'm happy to adjust my PR ;]

NekR commented:
Okay, trying to explain how to do it right :-)

* You can to commit source code (ES2015) too, not only generated code
* You have compiled code with custom installed `babel`, instead you should run `npm run install:build-deps` and after that you will be able to run `npm run build`
* We don't need `.babelrc` or other additional deps since all is already there
* Once you are able to build project with current deps, you most likely will have tests failing because of your changes (good if you don't :-)). Also you have to add tests for the new features (`npm test` to run tests)

Hope it helps a bit. Feel free to ask questions and sorry for a delay :-)

NekR commented:
Also there is a thing that I locked `v4` from new features because I started working on some breaking changes and that requires `v5`. Maybe we can add this to `v4` though, once it's done correctly.

domderen commented:
Hey,

Done and done ;) Fixed the build system stuff, removed the babel deps that I added, and used your build system. Also fixed the unit tests, and added new ones for this feature ;)

Thanks for the help with that.

Cheers,
Dom

NekR commented on tpls/runtime-template.js:
```diff
@@ -204,5 +204,22 @@ function applyUpdate(callback, errback) {
 }
 
+function updateOfflineService() {
+  if (navigator.serviceWorker) {
```
There is `hasSW()` function which should be used instead of `navigator.serviceWorker` check.

NekR commented on tpls/runtime-template.js:
```diff
@@ -204,5 +204,22 @@ function applyUpdate(callback, errback) {
+      return registration.update();
+    });
+  } else if (window.applicationCache) {
+    applicationCache.update();
```
I don't remember why, but I call application cache methods on iframe instance rather than on top instance. There probably should be a reason for that, so would be good to have it that way.

NekR commented on tpls/runtime-template.js:
```diff
@@ -204,5 +204,22 @@ function applyUpdate(callback, errback) {
+}
+
+<% if (typeof autoUpdate !== 'undefined' && autoUpdate === true) { %>
+  setInterval(function () { updateOfflineService(); }, <%- JSON.stringify(autoUpdateInterval) %>);
```
`<%- autoUpdateInterval %>` should be fine I guess (will generated number instead of string).

NekR commented:
@domderen hey, thank you for the changes. That's great indeed. But there are some things (which I missed in initial review) which would be good change and then I'll be able to merge it and release for `4.6.0` version.

* Items per review
* `package.json` doesn't need changes (it has just deps reorder)

Also we may have problems with multiple page instances opened and requested update many times. Not sure what's the good way here. Probably store some flag in `sessionStorage` and don't do duplicated updates.

domderen commented:
Nice, thanks for the review. I've fixed the points that you highlighted, and cleaned up `package.json` so it wouldn't show up in this PR. 

It's a good thought about handling multiple pages open, but with the default timeout of 1h, I think it wouldn't be too bothersome to have that update pinging from multiple tabs. So for the time being I didn't add anything like that, and can look at it in a separate PR. What do you think about that?

NekR commented:
@domderen Yeah, I think it's good to go now. Thank you very much 👍 

Regarding multiple tabs, yeah, separate PR for that is fine if you feel to do so.

NekR commented:
Thinking about it a bit more.. maybe it makes sense to just expose `runtime.update()` function then everyone will be able to do updates when they want and how, e.g. with interval as in this PR `setInterval(runtime.update, AN_HOUR)`.

domderen commented:
Sure that's doable as well, will have a look at it ;)

NekR commented:
@domderen I don't mean that you should do it, I'm just asking about your opinion about that :-)

domderen commented:
I like the freedom of that approach, which would allow me to define any arbitrary update logic. Maybe someone has a websocket notification when new version is available, and just wants to update the page on incoming message.

NekR commented:
@domderen okay, so you will be fine if I'll do that change and you will have to do `setInterval(runtime.update, AN_HOUR)` manually in your code? (which I want to do because I also like the freedom of that approach)

domderen commented:
Oh sorry, I meant that it would be cool to expose it on the runtime, but not necessarily remove it as it is right now. Current solution is simpler, when you just have to provide config and it works. But would additionally expose it on the runtime for people to call it manually if they have a more sophisticated usecase. Though personally I don't have such a requirements, and honestly I think that it would be rather a rare case.

NekR commented:
I see what you mean. I actually know some projects where they use websocket notifications to let webapp know about the update. But yeah, probably a rare case.

domderen commented:
Yeah, it's kinda like Docker's "batteries included but pluggable", where the simplest solution is provided out of the box with just config params, but if someone needs something more sophisticated, they can still do it.

NekR commented:
Hey @domderen, I just released `4.6.0` with this feature included. I a bit modified and it's now just `autoUpdate` option with ability to set `number`/`boolean`. Please checkout docs if something and thank you for contributing! 👍

zackify commented:
@NekR this now causes an error on build when using uglifyjs in webpack, since it isnt es5
<img width="701" alt="screen shot 2017-02-03 at 11 00 10 am" src="https://cloud.githubusercontent.com/assets/449136/22598022/fc6108c2-e9ff-11e6-9459-c0a7e0fcb0ea.png">
<img width="583" alt="screen shot 2017-02-03 at 11 00 19 am" src="https://cloud.githubusercontent.com/assets/449136/22598024/ff6e5056-e9ff-11e6-82d4-ba48e02dfd11.png">

NekR commented:
I know, see: https://github.com/NekR/offline-plugin/pull/177
I'll fix that myself soon.. sorry

NekR commented:
Okay, `4.6.1` is out, sorry for the trouble 😞

zackify commented:
@NekR no worries! Thanks for all the awesome work on this as always. And @domderen for the PR!
